File: 0878-nth-magical-number/0878-nth-magical-number.js
/**
 * @param {number} n
 * @param {number} a
 * @param {number} b
 * @return {number}
 */
var nthMagicalNumber = function(n, a, b) {
  const lcm = a * b / gcd(a, b);
  const mod = 1e9 + 7;
  let [lo, hi] = [2, 10e14];
  
  while (lo < hi) {
    const mid = lo + Math.floor((hi - lo) / 2);
    if (Math.floor(mid / a) + Math.floor(mid / b) - Math.floor(mid / lcm) < n) lo = mid + 1;
    else hi = mid;
  }
  
  return lo % mod;
};

const gcd = (a, b) => !b ? a : gcd(b, a % b);

File: 0878-nth-magical-number/0878-nth-magical-number.py
class Solution:
    def nthMagicalNumber(self, n: int, a: int, b: int) -> int:
        def gcd(a, b):
            if not b:
                return a
            return gcd(b, a % b)
        
        
        lcm = a * b / gcd(a, b)
        lo, hi = 2, 10**14
        while lo < hi:
            mid = (lo + hi) >> 1
            if mid // a + mid // b - mid // lcm < n:
                lo = mid + 1
            else:
                hi = mid
        return lo % (10**9 + 7)